### PR TITLE
replicate: add state switch

### PR DESCRIPTION
**Pull request by disksing**
Signed-off-by: disksing <i@disksing.com>

### What problem does this PR solve? 
Fix https://github.com/pingcap/pd/issues/2271

### What is changed and how it works?
- add a background job to check store down and switch states
- add test

### Check List 

Tests 
 - Unit test

Code changes
 - Has persistent data change

### CODE REVIEW

File: server/replicate/replicate.go [hunk 1]
```diff
@@ -118,13 +132,16 @@ func (m *ModeManager) drSwitchToSyncRecover() error {
 	defer m.Unlock()
 	id, err := m.idAlloc.Alloc()
 	if err != nil {
+		log.Warn("failed to switch to sync_recover state", zap.String("replicate-mode", "dr_async"), zap.Error(err))
```
Comment (shafreeck) [inline server/replicate/replicate.go:135]: Should be `dr_sync` here?

File: server/replicate/replicate.go [hunk 2]
```diff
@@ -107,9 +119,11 @@ func (m *ModeManager) drSwitchToAsync() error {
 	defer m.Unlock()
 	dr := drAutosyncStatus{State: drStateAsync}
 	if err := m.storage.SaveReplicateStatus(modeDRAutosync, dr); err != nil {
+		log.Warn("failed to switch to async state", zap.String("replicate-mode", "dr_async"), zap.Error(err))
 		return err
 	}
 	m.drAutosync = dr
+	log.Warn("switched to async state", zap.String("replicate-mode", "dr_async"))
```
Comment (shafreeck) [inline server/replicate/replicate.go:126]: Use `log.Info`?

File: server/replicate/replicate.go [hunk 3]
```diff
@@ -118,13 +132,16 @@ func (m *ModeManager) drSwitchToSyncRecover() error {
...
 		return err
 	}
-	dr := drAutosyncStatus{State: drStateSyncRecover, RecoverID: id}
+	dr := drAutosyncStatus{State: drStateSyncRecover, RecoverID: id, RecoverStartTime: time.Now()}
 	if err = m.storage.SaveReplicateStatus(modeDRAutosync, dr); err != nil {
+		log.Warn("failed to switch to sync_recover state", zap.String("replicate-mode", "dr_async"), zap.Error(err))
 		return err
 	}
 	m.drAutosync = dr
+	log.Warn("switched to sync_recover state", zap.String("replicate-mode", "dr_async"))
```
Comment (shafreeck) [inline server/replicate/replicate.go:144]: Use `log.Info`?

File: server/replicate/replicate.go [hunk 4]
```diff
@@ -133,8 +150,90 @@ func (m *ModeManager) drSwitchToSync() error {
 	defer m.Unlock()
 	dr := drAutosyncStatus{State: drStateSync}
 	if err := m.storage.SaveReplicateStatus(modeDRAutosync, dr); err != nil {
+		log.Warn("failed to switch to sync state", zap.String("replicate-mode", "dr_async"), zap.Error(err))
```
Comment (shafreeck) [inline server/replicate/replicate.go:153]: `dr_async` -> `dr_sync`

File: server/replicate/replicate.go [hunk 5]
```diff
@@ -133,8 +150,90 @@ func (m *ModeManager) drSwitchToSync() error {
...
 		return err
 	}
 	m.drAutosync = dr
+	log.Warn("switched to sync state", zap.String("replicate-mode", "dr_async"))
```
Comment (shafreeck) [inline server/replicate/replicate.go:157]: diito

File: server/replicate/replicate.go [hunk 6]
```diff
@@ -133,8 +150,90 @@ func (m *ModeManager) drSwitchToSync() error {
...
+
+const (
+	idleTimeout  = time.Minute
+	tickInterval = time.Second * 10
+)
+
+// Run starts the background job.
+func (m *ModeManager) Run(quit chan struct{}) {
+	select {
+	case <-time.After(idleTimeout):
```
Comment (nolouch) [inline server/replicate/replicate.go:175]: Why need to wait for an idle timeout?  better to add a comment.

Comment (disksing) [reply server/replicate/replicate.go:175]: ok, good idea

**Pull request merged**